test/extensions/filters/http/common/jwks_fetcher_test.cc

```diff
@@ -0,0 +1,160 @@
+#include <chrono>
+#include <thread>
+
+#include "common/http/message_impl.h"
+#include "common/protobuf/utility.h"
+
+#include "extensions/filters/http/common/jwks_fetcher.h"
+
+#include "test/extensions/filters/http/common/test_common.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/server/mocks.h"
+#include "test/test_common/utility.h"
+
+using ::envoy::api::v2::core::HttpUri;
+using ::google::jwt_verify::Status;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Common {
+namespace {
+
+const std::string JwksUri = R"(
+uri: https://pubkey_server/pubkey_path
+cluster: pubkey_cluster
+timeout:
+  seconds: 5
+)";
+
+class JwksFetcherTest : public ::testing::Test {
+public:
+  void SetUp() { MessageUtil::loadFromYaml(JwksUri, uri_); }
+  HttpUri uri_;
+  testing::NiceMock<Server::Configuration::MockFactoryContext> mock_factory_ctx_;
+};
+
+// A mock HTTP upstream with response body.
+class MockUpstream {
+public:
+  MockUpstream(Upstream::MockClusterManager& mock_cm, const std::string& status,
+               const std::string& response_body)
+      : request_(&mock_cm.async_client_), status_(status), response_body_(response_body) {
+    ON_CALL(mock_cm.async_client_, send_(testing::_, testing::_, testing::_))
+        .WillByDefault(testing::Invoke([this](Http::MessagePtr&, Http::AsyncClient::Callbacks& cb,
+                                              const absl::optional<std::chrono::milliseconds>&)
+                                           -> Http::AsyncClient::Request* {
+          Http::MessagePtr response_message(new Http::ResponseMessageImpl(
+              Http::HeaderMapPtr{new Http::TestHeaderMapImpl{{":status", status_}}}));
+          if (response_body_.length()) {
+            response_message->body().reset(new Buffer::OwnedImpl(response_body_));
+          } else {
+            response_message->body().reset(nullptr);
+          }
+          cb.onSuccess(std::move(response_message));
+          return &request_;
+        }));
+  }
+
+  MockUpstream(Upstream::MockClusterManager& mock_cm, Http::MockAsyncClientRequest* request)
+      : request_(&mock_cm.async_client_) {
+    ON_CALL(mock_cm.async_client_, send_(testing::_, testing::_, testing::_))
+        .WillByDefault(testing::Invoke([request](Http::MessagePtr&, Http::AsyncClient::Callbacks&,
```

Comment:
WillByDefault(testing::Return(request));

Reply:
Done in 6ca0471